Deploy-CAM.ps1

```diff
@@ -2924,58 +2924,101 @@ else {
+        }
 
-            $domainAdminCredential = Get-Credential -Message $domainAdminMessage
-            $confirmedPassword = Read-Host -AsSecureString "Please re-enter the password"
-
-            # Need plaintext password to check if same
-            $clearPassword = ConvertTo-Plaintext $confirmedPassword
-            if (-not ($domainAdminCredential.GetNetworkCredential().Password -ceq $clearPassword)) {
-                # don't match- try again.
-                Write-Host "The entered passwords do not match."
-                $domainAdminCredential = $null
+        # only check if it is not deployOverDC
+        if (-not $deployOverDC) {
+            if ((-not ($username -match '^[A-Za-z0-9]+(?:[_-][A-Za-z0-9]+)*$')) -or ($username.Length -gt 20)) {
+                Write-Host "`nPlease enter a valid username. It can only contain letters and numbers and cannot be longer than 20 characters." -ForegroundColor Red -BackgroundColor Yellow
+                $username = $null
                 continue
             }
+            # reserved usernames
+            # https://docs.microsoft.com/en-us/azure/virtual-machines/windows/faq
+            $reservedUsername = @('administrator', 'admin', 'user', 'user1', 'test', 'user2', 'test1', 'user3', 'admin1', '1', '123', 'a',
+                                  'actuser', 'adm', 'admin2', 'aspnet', 'backup', 'console', 'david', 'guest', 'john', 'owner', 'root', 'server',
+                                  'sql', 'support', 'support_388945a0', 'sys', 'test2', 'test3', 'user4', 'user5' )
+            
+            if ($username -in $reservedUsername) {
+                Write-Host "`n$username is a reserved username. Please try again" -ForegroundColor Red -BackgroundColor Yellow
+                $username = $null
+                continue                
+            }
         }
 
-        if ((-not ($domainAdminCredential.UserName -imatch '\w+')) -or ($domainAdminCredential.Username.Length -gt 20)) {
-            Write-Host "Please enter a valid username. It can only contain letters and numbers and cannot be longer than 20 characters."
-            $domainAdminCredential = $null
-            continue
+    } while ( -not $username )
+
+    # Password
+    $password = $null
+    $firstRun = 1
+    $prompted = $null
+    do {
+        if ($domainAdminCredential -and $firstRun) {
+            $password = $domainAdminCredential.GetNetworkCredential().Password
+            $firstRun = $null
+        }
+
+        if ( -not $password ) {
+            $psw = Read-Host -AsSecureString "Please enter the password"
+            $password = ConvertTo-Plaintext $psw
+            $prompted = 1
+        }
+
+        # Don't check password if deploying over DC since in that case it's the DC's password complexity rules.
+        if (-not $deployOverDC ) {
+            if (-not ($password -match '^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[$@$!%*?&])[A-Za-z\d$@$!%*?&]{12,}')) {
```

Comment:
I suspect that attempting to enforce password composition rules in a regex is an unwinnable fight... it will certainly be an unreadable one. :) I tried this regex against several strings that would be perfectly fine passwords elsewhere (generated via password generators and phrases including whitespace) but are rejected by this regex.

If we are going to enforce composition requirements, I think we'd be better off writing a method to handle it than attempting to use a regex.

Reply:
I will remove the checking for password complexity to only check for 12 characters